packages/phpstan-rules/src/Rules/ForbiddenAssignInLoopRule.php

```diff
@@ -0,0 +1,182 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Symplify\PHPStanRules\Rules;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\Assign;
+use PhpParser\Node\Expr\Variable;
+use PhpParser\Node\Stmt\Do_;
+use PhpParser\Node\Stmt\For_;
+use PhpParser\Node\Stmt\Foreach_;
+use PhpParser\Node\Stmt\While_;
+use PhpParser\NodeFinder;
+use PHPStan\Analyser\Scope;
+use Symplify\Astral\Naming\SimpleNameResolver;
+use Symplify\PHPStanRules\ValueObject\PHPStanAttributeKey;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @see \Symplify\PHPStanRules\Tests\Rules\ForbiddenAssignInLoopRule\ForbiddenAssignInLoopRuleTest
+ */
+final class ForbiddenAssignInLoopRule extends AbstractSymplifyRule
+{
+    /**
+     * @var string
+     */
+    public const ERROR_MESSAGE = 'Assign in loop is not allowed.';
+
+    /**
+     * @var array<string, array<int, string>>
+     */
+    private const LOOP_STMTS_CHECKS = [
+        Do_::class => ['cond'],
+        For_::class => ['init', 'cond', 'loop'],
+        Foreach_::class => ['expr', 'keyVar', 'valueVar'],
+        While_::class => ['cond'],
+    ];
+
+    /**
+     * @var NodeFinder
+     */
+    private $nodeFinder;
+
+    /**
+     * @var SimpleNameResolver
+     */
+    private $simpleNameResolver;
+
+    public function __construct(NodeFinder $nodeFinder, SimpleNameResolver $simpleNameResolver)
+    {
+        $this->nodeFinder = $nodeFinder;
+        $this->simpleNameResolver = $simpleNameResolver;
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        return [Do_::class, For_::class, Foreach_::class, While_::class];
+    }
+
+    /**
+     * @param Do_|For_|Foreach_|While_ $node
+     * @return string[]
+     */
+    public function process(Node $node, Scope $scope): array
+    {
+        /** @var Assign[] $assigns */
+        $assigns = $this->nodeFinder->findInstanceOf($node->stmts, Assign::class);
+        if ($assigns === []) {
+            return [];
+        }
+
+        $nodeClass = get_class($node);
+        foreach (self::LOOP_STMTS_CHECKS[$nodeClass] as $expr) {
+            /** @var Variable[] $variables */
+            $variables = $this->nodeFinder->find($node->{$expr}, function (Node $n): bool {
```

Comment:
This magic looks dangerous. What if $expr value changes in original nodes?

What about using the node directly?

```suggestion
            $variables = $this->nodeFinder->find($node, function (Node $n): bool {
```

Reply:
There are many keys to check:

```php
    private const LOOP_STMTS_CHECKS = [
        Do_::class => ['cond'],
        For_::class => ['init', 'cond', 'loop'],
        Foreach_::class => ['expr', 'keyVar', 'valueVar'],
        While_::class => ['cond'],
    ];
```
is that efficient, need to create 4 methods at least and 1 method caller for that to check instanceof each.

Reply:
It's less code to write now, but magical variables show as impossible to refactor.

If `cond` will be renamed to `condition`, Rector cannot upgrade it. Neither PHPStan undrestand this.


Better create 4 methods, one per type.

Also, to get instanceof specific type, `$nodeFinder` has `findInstanceOf()` method

Reply:
I will try.

Reply:
implemented.